src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -117908,5 +117520,1523 @@ class Program
+        }
+
+        [Fact]
+        [WorkItem(38638, "https://github.com/dotnet/roslyn/issues/38638")]
+        public void MaybeNullT_14()
+        {
+            var source =
+@"#nullable enable
+using System.Diagnostics.CodeAnalysis;
+class C<T>
+{
+    T F1 = default!;
+    [AllowNull]T F2 = default!;
+    [MaybeNull]T F3 = default!;
+    void M1(T x, [AllowNull]T y)
+    {
+        F1 = x;
+        F2 = x;
+        F3 = x;
+        F1 = y; // 1
+        F2 = y;
+        F3 = y; // 2
+    }
+}";
+            var comp = CreateCompilation(new[] { source, AllowNullAttributeDefinition, MaybeNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (13,14): warning CS8601: Possible null reference assignment.
+                //         F1 = y; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "y").WithLocation(13, 14),
+                // (15,14): warning CS8601: Possible null reference assignment.
+                //         F3 = y; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "y").WithLocation(15, 14));
+        }
+
+        [Fact]
+        [WorkItem(38638, "https://github.com/dotnet/roslyn/issues/38638")]
+        public void MaybeNullT_15()
+        {
+            var source =
+@"#nullable enable
+using System.Diagnostics.CodeAnalysis;
+class Program
+{
+    static T F1<T, U>(bool b, T t, U u) where U : T => b ? t : u;
+    static T F2<T, U>(bool b, T t, [AllowNull]U u) where U : T => b ? t : u;
+    static T F3<T, U>(bool b, [AllowNull]T t, U u) where U : T => b ? t : u;
+    static T F4<T, U>(bool b, [AllowNull]T t, [AllowNull]U u) where U : T => b ? t : u;
+}";
+            var comp = CreateCompilation(new[] { source, AllowNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (6,67): warning CS8603: Possible null reference return.
+                //     static T F2<T, U>(bool b, T t, [AllowNull]U u) where U : T => b ? t : u;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "b ? t : u").WithLocation(6, 67),
+                // (7,67): warning CS8603: Possible null reference return.
+                //     static T F3<T, U>(bool b, [AllowNull]T t, U u) where U : T => b ? t : u;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "b ? t : u").WithLocation(7, 67),
+                // (8,78): warning CS8603: Possible null reference return.
+                //     static T F4<T, U>(bool b, [AllowNull]T t, [AllowNull]U u) where U : T => b ? t : u;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "b ? t : u").WithLocation(8, 78));
+        }
+
+        [Fact]
+        [WorkItem(38638, "https://github.com/dotnet/roslyn/issues/38638")]
+        public void MaybeNullT_16()
+        {
+            var source =
+@"#nullable enable
+using System.Diagnostics.CodeAnalysis;
+class Program
+{
+    static T F0<T>(T t) => t ??= default;
```

Comment:
This appears to still be using ??=

Reply:
Ah, sorry, I'd missed that you suggested `??` rather than `??=`. I'll add a test for `??` in a follow up PR.

---
In reply to: [348736052](https://github.com/dotnet/roslyn/pull/39778#discussion_r348736052) [](ancestors = 348736052)